application/forms/EntryForm.php

```diff
@@ -311,114 +317,233 @@ protected function assemble()
+        $membershipHash = $this->getMembershipHashFromRecipients($scheduleId);
         $recipients = array_map(function ($recipient) {
             return explode(':', $recipient, 2);
         }, explode(',', $this->getValue('recipient')));
 
         $db = Database::get();
         $db->beginTransaction();
 
-        $db->insert('timeperiod', ['owned_by_schedule_id' => $scheduleId]);
-        $timeperiodId = $db->lastInsertId();
+        $members = ScheduleMember::on(Database::get())
+            ->with('timeperiod')
+            ->filter(
+                Filter::all(
+                    Filter::equal('timeperiod.owned_by_schedule_id', $scheduleId),
+                    Filter::equal('schedule_member.membership_hash', $membershipHash)
+                )
+            );
+
+        if ($members->count() > 0) {
+            $timeperiodId = $members->first()->timeperiod_id;
+        } else {
+            $db->insert('timeperiod', ['owned_by_schedule_id' => $scheduleId]);
+            $timeperiodId = $db->lastInsertId();
+        }
 
         $db->insert('timeperiod_entry', $data + ['timeperiod_id' => $timeperiodId]);
 
-        foreach ($recipients as list($type, $id)) {
-            if ($type === 'contact') {
-                $db->insert('schedule_member', [
-                    'schedule_id' => $scheduleId,
-                    'timeperiod_id' => $timeperiodId,
-                    'contact_id' => $id
-                ]);
-            } elseif ($type === 'group') {
-                $db->insert('schedule_member', [
-                    'schedule_id' => $scheduleId,
-                    'timeperiod_id' => $timeperiodId,
-                    'contactgroup_id' => $id
-                ]);
+        if ($members->count() === 0) {
+            $binaryMemberHash = (new Binary([]))->toDb($membershipHash, null, null);
+            foreach ($recipients as list($type, $id)) {
+                if ($type === 'contact') {
+                    $db->insert('schedule_member', [
+                        'schedule_id'     => $scheduleId,
+                        'timeperiod_id'   => $timeperiodId,
+                        'contact_id'      => $id,
+                        'membership_hash' => $binaryMemberHash
+                    ]);
+                } elseif ($type === 'group') {
+                    $db->insert('schedule_member', [
+                        'schedule_id'     => $scheduleId,
+                        'timeperiod_id'   => $timeperiodId,
+                        'contactgroup_id' => $id,
+                        'membership_hash' => $binaryMemberHash
+                    ]);
+                }
             }
         }
 
         $db->commitTransaction();
     }
 
+    /**
+     * Get membership hash for the recipients in timeperiod entry for the given schedule
+     *
+     * @param int $scheduleID
+     *
+     * @return string
+     */
+    public function getMembershipHashFromRecipients(int $scheduleID): string
```

Comment:
Please use `$scheduleId` (with a lowercase `d` at the end) instead. `addEntry` does it already.